providers/common/ciphers/aes_gcm.c

```diff
@@ -0,0 +1,568 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/crypto.h>
+#include <openssl/core_numbers.h>
+#include <openssl/core_names.h>
+#include <openssl/evp.h>
+#include <openssl/params.h>
+#include "internal/rand_int.h"
+#include "internal/cryptlib.h"
+#include "internal/provider_algs.h"
+#include "internal/providercommonerr.h"
+#include "ciphers_locl.h"
+
+#define AEAD_GCM_FLAGS (EVP_CIPH_FLAG_AEAD_CIPHER | EVP_CIPH_FLAG_DEFAULT_ASN1 \
+                       | EVP_CIPH_CUSTOM_IV | EVP_CIPH_FLAG_CUSTOM_CIPHER      \
+                       | EVP_CIPH_ALWAYS_CALL_INIT | EVP_CIPH_CTRL_INIT        \
+                       | EVP_CIPH_CUSTOM_COPY)
+
+static OSSL_OP_cipher_dupctx_fn aes_gcm_dupctx;
+static OSSL_OP_cipher_freectx_fn aes_gcm_freectx;
+static OSSL_OP_cipher_encrypt_init_fn aes_gcm_einit;
+static OSSL_OP_cipher_decrypt_init_fn aes_gcm_dinit;
+static OSSL_OP_cipher_ctx_get_params_fn aes_gcm_ctx_get_params;
+static OSSL_OP_cipher_ctx_set_params_fn aes_gcm_ctx_set_params;
+static OSSL_OP_cipher_cipher_fn aes_gcm_cipher;
+static OSSL_OP_cipher_update_fn aes_gcm_stream_update;
+static OSSL_OP_cipher_final_fn aes_gcm_stream_final;
+
+static int aes_gcm_tls_init(PROV_AES_GCM_CTX *dat, unsigned char *aad,
+                            size_t aad_len);
+static int aes_gcm_tls_iv_set_fixed(PROV_AES_GCM_CTX *ctx, unsigned char *iv,
+                                    size_t len);
+static int aes_gcm_tls_cipher(PROV_AES_GCM_CTX *ctx,
+                              unsigned char *out, size_t *padlen,
+                              const unsigned char *in, size_t len);
+static int aes_gcm_cipher_internal(PROV_AES_GCM_CTX *ctx, unsigned char *out,
+                                   size_t *padlen, const unsigned char *in,
+                                   size_t len);
+
+static void *aes_gcm_newctx(void *provctx, size_t keybits)
+{
+    PROV_AES_GCM_CTX *ctx = OPENSSL_zalloc(sizeof(*ctx));
+
+    ctx->pad = 1;
+    ctx->mode = EVP_CIPH_GCM_MODE;
+    ctx->taglen = -1;
+    ctx->tls_aad_len = -1;
+    ctx->ivlen = (EVP_GCM_TLS_FIXED_IV_LEN + EVP_GCM_TLS_EXPLICIT_IV_LEN);
+    ctx->keylen = keybits / 8;
+    ctx->hw = PROV_AES_HW_gcm(keybits);
+    ctx->libctx = provctx;
```

Comment:
I think we're supposed to use the macro `PROV_LIBRARY_CONTEXT_OF` here, but I forget why (because it currently doesn't do anything). Perhaps @levitte can remind me.

Reply:
Updated - maybe it will do something at some point if the provctx is more than just a libctx at some point

Reply:
Exactly